File: src/praisonai/praisonai/cli/features/slash_commands.py
"""
Slash Commands System for PraisonAI CLI.

Inspired by Aider's command system and Gemini CLI's slash commands.
Provides interactive commands like /help, /cost, /plan, /model, /clear, etc.

Architecture:
- SlashCommand: Base class for all commands
- SlashCommandRegistry: Manages command registration and lookup
- SlashCommandParser: Parses user input for slash commands
"""

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class CommandKind(Enum):
    """Type of slash command."""
    BUILT_IN = "built_in"
    CUSTOM = "custom"
    MCP = "mcp"


@dataclass
class SlashCommand:
    """
    Represents a slash command.
    
    Attributes:
        name: Primary command name (e.g., "help", "cost")
        description: Short description for help text
        action: Callable that executes the command
        alt_names: Alternative names/aliases
        sub_commands: Nested sub-commands
        kind: Type of command (built-in, custom, MCP)
        auto_execute: Whether to execute without confirmation
    """
    name: str
    description: str
    action: Optional[Callable[["CommandContext", str], Any]] = None
    alt_names: List[str] = field(default_factory=list)
    sub_commands: List["SlashCommand"] = field(default_factory=list)
    kind: CommandKind = CommandKind.BUILT_IN
    auto_execute: bool = True
    
    def __post_init__(self):
        """Validate command after initialization."""
        if not self.name:
            raise ValueError("Command name cannot be empty")
        if not self.name.isalnum() and "_" not in self.name and "-" not in self.name:
            raise ValueError(f"Invalid command name: {self.name}")


@dataclass
class CommandContext:
    """
    Context passed to command actions.
    
    Contains all services and state needed by commands.
    """
    # Core services
    agent: Any = None  # PraisonAI Agent instance
    config: Dict[str, Any] = field(default_factory=dict)
    
    # Session state
    session_id: Optional[str] = None
    session_start_time: Optional[float] = None
    
    # Metrics
    total_tokens: int = 0
    total_cost: float = 0.0
    prompt_count: int = 0
    
    # UI callbacks
    print_fn: Callable[[str], None] = print
    input_fn: Callable[[str], str] = input
    
    # Command invocation info
    raw_input: str = ""
    command_name: str = ""
    args: str = ""


@dataclass
class ParsedSlashCommand:
    """Result of parsing a slash command."""
    command: Optional[SlashCommand] = None
    args: str = ""
    canonical_path: List[str] = field(default_factory=list)
    error: Optional[str] = None


class SlashCommandRegistry:
    """
    Registry for slash commands.
    
    Manages command registration, lookup, and help generation.
    """
    
    def __init__(self):
        self._commands: Dict[str, SlashCommand] = {}
        self._aliases: Dict[str, str] = {}  # alias -> primary name
    
    def register(self, command: SlashCommand) -> None:
        """Register a slash command."""
        if command.name in self._commands:
            logger.warning(f"Overwriting existing command: {command.name}")
        
        self._commands[command.name] = command
        
        # Register aliases
        for alias in command.alt_names:
            if alias in self._aliases:
                logger.warning(f"Alias '{alias}' already registered")
            self._aliases[alias] = command.name
    
    def unregister(self, name: str) -> bool:
        """Unregister a command by name."""
        if name in self._commands:
            cmd = self._commands.pop(name)
            # Remove aliases
            for alias in cmd.alt_names:
                self._aliases.pop(alias, None)
            return True
        return False
    
    def get(self, name: str) -> Optional[SlashCommand]:
        """Get a command by name or alias."""
        # Check primary name first
        if name in self._commands:
            return self._commands[name]
        # Check aliases
        if name in self._aliases:
            return self._commands.get(self._aliases[name])
        return None
    
    def get_all(self) -> List[SlashCommand]:
        """Get all registered commands."""
        return list(self._commands.values())
    
    def get_names(self) -> List[str]:
        """Get all command names (including aliases)."""
        names = list(self._commands.keys())
        names.extend(self._aliases.keys())
        return sorted(set(names))


class SlashCommandParser:
    """
    Parser for slash commands.
    
    Handles parsing of user input into command and arguments.
    """
    
    def __init__(self, registry: SlashCommandRegistry):
        self.registry = registry
    
    def is_slash_command(self, query: str) -> bool:
        """Check if input is a slash command."""
        query = query.strip()
        if not query.startswith('/'):
            return False
        # Exclude code comments
        if query.startswith('//') or query.startswith('/*'):
            return False
        return True
    
    def parse(self, query: str) -> ParsedSlashCommand:
        """
        Parse a slash command string.
        
        Args:
            query: Raw input string (e.g., "/help model" or "/cost")
            
        Returns:
            ParsedSlashCommand with resolved command and arguments
        """
        if not self.is_slash_command(query):
            return ParsedSlashCommand(error="Not a slash command")
        
        trimmed = query.strip()
        parts = trimmed[1:].strip().split()  # Remove leading /
        
        if not parts:
            return ParsedSlashCommand(error="Empty command")
        
        command_path = parts
        canonical_path: List[str] = []
        current_commands = self.registry.get_all()
        command_to_execute: Optional[SlashCommand] = None
        path_index = 0
        
        for part in command_path:
            # First pass: exact match on primary name
            found_command = None
            for cmd in current_commands:
                if cmd.name == part:
                    found_command = cmd
                    break
            
            # Second pass: check aliases
            if not found_command:
                for cmd in current_commands:
                    if part in cmd.alt_names:
                        found_command = cmd
                        break
            
            if found_command:
                command_to_execute = found_command
                canonical_path.append(found_command.name)
                path_index += 1
                if found_command.sub_commands:
                    current_commands = found_command.sub_commands
                else:
                    break
            else:
                break
        
        args = ' '.join(parts[path_index:])
        
        return ParsedSlashCommand(
            command=command_to_execute,
            args=args,
            canonical_path=canonical_path
        )


# ============================================================================
# Built-in Commands
# ============================================================================

def cmd_help(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show help for commands."""
    from rich.console import Console
    from rich.table import Table
    
    console = Console()
    
    if args:
        # Help for specific command
        registry = context.config.get("command_registry")
        if registry:
            cmd = registry.get(args)
            if cmd:
                console.print(f"\n[bold cyan]/{cmd.name}[/bold cyan]")
                console.print(f"  {cmd.description}")
                if cmd.alt_names:
                    console.print(f"  Aliases: {', '.join(cmd.alt_names)}")
                if cmd.sub_commands:
                    console.print("\n  Sub-commands:")
                    for sub in cmd.sub_commands:
                        console.print(f"    /{cmd.name} {sub.name} - {sub.description}")
                return {"type": "help", "command": cmd.name}
            else:
                console.print(f"[red]Unknown command: {args}[/red]")
                return {"type": "error", "message": f"Unknown command: {args}"}
    
    # General help
    table = Table(title="Available Commands", show_header=True)
    table.add_column("Command", style="cyan")
    table.add_column("Description")
    
    registry = context.config.get("command_registry")
    if registry:
        for cmd in registry.get_all():
            aliases = f" ({', '.join(cmd.alt_names)})" if cmd.alt_names else ""
            table.add_row(f"/{cmd.name}{aliases}", cmd.description)
    
    console.print(table)
    return {"type": "help"}


def cmd_cost(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show session cost and token usage."""
    from rich.console import Console
    from rich.panel import Panel
    
    console = Console()
    
    cost_info = f"""
[bold]Session Statistics[/bold]

Total Tokens: {context.total_tokens:,}
Total Cost: ${context.total_cost:.4f}
Prompts: {context.prompt_count}
"""
    
    if context.session_start_time:
        import time
        duration = time.time() - context.session_start_time
        minutes = int(duration // 60)
        seconds = int(duration % 60)
        cost_info += f"Duration: {minutes}m {seconds}s\n"
    
    console.print(Panel(cost_info, title="💰 Cost Tracker", border_style="green"))
    
    return {
        "type": "stats",
        "tokens": context.total_tokens,
        "cost": context.total_cost,
        "prompts": context.prompt_count
    }


def cmd_clear(context: CommandContext, args: str) -> Dict[str, Any]:
    """Clear the conversation history."""
    from rich.console import Console
    console = Console()
    
    # Reset agent conversation if available
    if context.agent and hasattr(context.agent, 'clear_history'):
        context.agent.clear_history()
    
    console.print("[green]✓ Conversation cleared[/green]")
    return {"type": "clear"}


def cmd_model(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show or change the current model."""
    from rich.console import Console
    console = Console()
    
    if args:
        # Change model
        if context.agent and hasattr(context.agent, 'set_model'):
            context.agent.set_model(args)
            console.print(f"[green]✓ Model changed to: {args}[/green]")
            return {"type": "model_change", "model": args}
        else:
            console.print(f"[yellow]Model change requested: {args}[/yellow]")
            return {"type": "model_change", "model": args}
    else:
        # Show current model
        current_model = "unknown"
        if context.agent and hasattr(context.agent, 'llm'):
            current_model = getattr(context.agent.llm, 'model', 'unknown')
        console.print(f"[cyan]Current model: {current_model}[/cyan]")
        return {"type": "model_info", "model": current_model}


def cmd_tokens(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show token usage breakdown."""
    from rich.console import Console
    from rich.table import Table
    
    console = Console()
    
    table = Table(title="Token Usage", show_header=True)
    table.add_column("Metric", style="cyan")
    table.add_column("Value", justify="right")
    
    table.add_row("Total Tokens", f"{context.total_tokens:,}")
    table.add_row("Prompts Sent", str(context.prompt_count))
    if context.prompt_count > 0:
        avg = context.total_tokens / context.prompt_count
        table.add_row("Avg Tokens/Prompt", f"{avg:.0f}")
    
    console.print(table)
    return {"type": "tokens", "total": context.total_tokens}


def cmd_plan(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show or create an execution plan."""
    from rich.console import Console
    console = Console()
    
    if args:
        console.print(f"[cyan]Creating plan for: {args}[/cyan]")
        # Trigger planning mode
        return {"type": "submit_prompt", "content": f"Create a detailed plan for: {args}"}
    else:
        console.print("[yellow]Usage: /plan <task description>[/yellow]")
        return {"type": "help", "command": "plan"}


def cmd_undo(context: CommandContext, args: str) -> Dict[str, Any]:
    """Undo the last change."""
    from rich.console import Console
    console = Console()
    
    # Check for git integration
    try:
        import subprocess
        result = subprocess.run(
            ["git", "diff", "--stat", "HEAD~1"],
            capture_output=True,
            text=True,
            cwd="."
        )
        if result.returncode == 0:
            console.print("[yellow]Last commit changes:[/yellow]")
            console.print(result.stdout)
            console.print("\n[cyan]Run 'git reset --soft HEAD~1' to undo[/cyan]")
        else:
            console.print("[red]No git history available[/red]")
    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
    
    return {"type": "undo"}


def cmd_diff(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show git diff of changes."""
    from rich.console import Console
    from rich.syntax import Syntax
    
    console = Console()
    
    try:
        import subprocess
        result = subprocess.run(
            ["git", "diff"],
            capture_output=True,
            text=True,
            cwd="."
        )
        if result.stdout:
            syntax = Syntax(result.stdout, "diff", theme="monokai")
            console.print(syntax)
        else:
            console.print("[green]No uncommitted changes[/green]")
    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
    
    return {"type": "diff"}


def cmd_commit(context: CommandContext, args: str) -> Dict[str, Any]:
    """Commit changes with AI-generated message."""
    from rich.console import Console
    console = Console()
    
    console.print("[cyan]Generating commit message...[/cyan]")
    
    # This will trigger the commit command handler
    return {"type": "commit", "message": args if args else None}


def cmd_exit(context: CommandContext, args: str) -> Dict[str, Any]:
    """Exit the CLI."""
    from rich.console import Console
    console = Console()
    
    console.print("[yellow]Goodbye![/yellow]")
    return {"type": "exit"}


def cmd_settings(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show current settings."""
    from rich.console import Console
    from rich.table import Table
    
    console = Console()
    
    table = Table(title="Current Settings", show_header=True)
    table.add_column("Setting", style="cyan")
    table.add_column("Value")
    
    for key, value in context.config.items():
        if key != "command_registry":  # Skip internal objects
            table.add_row(key, str(value))
    
    console.print(table)
    return {"type": "settings"}


def cmd_map(context: CommandContext, args: str) -> Dict[str, Any]:
    """Show repository map."""
    from rich.console import Console
    console = Console()
    
    console.print("[cyan]Generating repository map...[/cyan]")
    # This will be implemented with RepoMap feature
    return {"type": "map"}


# ============================================================================
# Registry Setup
# ============================================================================

def create_default_registry() -> SlashCommandRegistry:
    """Create registry with default built-in commands."""
    registry = SlashCommandRegistry()
    
    # Core commands
    registry.register(SlashCommand(
        name="help",
        description="Show help for commands",
        action=cmd_help,
        alt_names=["h", "?"]
    ))
    
    registry.register(SlashCommand(
        name="cost",
        description="Show session cost and token usage",
        action=cmd_cost,
        alt_names=["usage", "stats"]
    ))
    
    registry.register(SlashCommand(
        name="clear",
        description="Clear conversation history",
        action=cmd_clear,
        alt_names=["reset"]
    ))
    
    registry.register(SlashCommand(
        name="model",
        description="Show or change the current model",
        action=cmd_model,
        alt_names=["m"]
    ))
    
    registry.register(SlashCommand(
        name="tokens",
        description="Show token usage breakdown",
        action=cmd_tokens
    ))
    
    registry.register(SlashCommand(
        name="plan",
        description="Create an execution plan for a task",
        action=cmd_plan
    ))
    
    registry.register(SlashCommand(
        name="undo",
        description="Undo the last change",
        action=cmd_undo
    ))
    
    registry.register(SlashCommand(
        name="diff",
        description="Show git diff of changes",
        action=cmd_diff
    ))
    
    registry.register(SlashCommand(
        name="commit",
        description="Commit changes with AI-generated message",
        action=cmd_commit
    ))
    
    registry.register(SlashCommand(
        name="exit",
        description="Exit the CLI",
        action=cmd_exit,
        alt_names=["quit", "q"]
    ))
    
    registry.register(SlashCommand(
        name="settings",
        description="Show current settings",
        action=cmd_settings
    ))
    
    registry.register(SlashCommand(
        name="map",
        description="Show repository map",
        action=cmd_map,
        alt_names=["repo"]
    ))
    
    return registry


# ============================================================================
# Handler for CLI Integration
# ============================================================================

class SlashCommandHandler:
    """
    Handler for integrating slash commands with PraisonAI CLI.
    """
    
    def __init__(self, verbose: bool = False):
        self.verbose = verbose
        self.registry = create_default_registry()
        self.parser = SlashCommandParser(self.registry)
        self._context: Optional[CommandContext] = None
    
    def set_context(self, context: CommandContext) -> None:
        """Set the command context."""
        self._context = context
        context.config["command_registry"] = self.registry
    
    def is_command(self, query: str) -> bool:
        """Check if input is a slash command."""
        return self.parser.is_slash_command(query)
    
    def execute(self, query: str) -> Optional[Dict[str, Any]]:
        """
        Execute a slash command.
        
        Returns:
            Command result dict or None if not a command
        """
        if not self.is_command(query):
            return None
        
        parsed = self.parser.parse(query)
        
        if parsed.error:
            if self.verbose:
                print(f"[red]Error: {parsed.error}[/red]")
            return {"type": "error", "message": parsed.error}
        
        if not parsed.command:
            return {"type": "error", "message": "Unknown command"}
        
        if not parsed.command.action:
            return {"type": "error", "message": f"Command '{parsed.command.name}' has no action"}
        
        # Create context if not set
        if not self._context:
            self._context = CommandContext()
            self._context.config["command_registry"] = self.registry
        
        self._context.raw_input = query
        self._context.command_name = parsed.command.name
        self._context.args = parsed.args
        
        try:
            result = parsed.command.action(self._context, parsed.args)
            return result
        except Exception as e:
            logger.error(f"Error executing command: {e}")
            return {"type": "error", "message": str(e)}
    
    def register_command(self, command: SlashCommand) -> None:
        """Register a custom command."""
        self.registry.register(command)
    
    def get_completions(self, partial: str) -> List[str]:
        """Get command completions for partial input."""
        if not partial.startswith('/'):
            return []
        
        partial_cmd = partial[1:].lower()
        completions = []
        
        for name in self.registry.get_names():
            if name.lower().startswith(partial_cmd):
                completions.append(f"/{name}")
        
        return sorted(completions)

File: src/praisonai-ts/src/cli/features/slash-commands.ts
/**
 * Slash Commands - Interactive command system for CLI
 * Supports /help, /cost, /clear, /model, /tokens, /plan, /undo, /diff, /commit, /exit, /settings, /map
 */

export interface SlashCommand {
  name: string;
  aliases?: string[];
  description: string;
  usage?: string;
  execute: (args: string[], context: SlashCommandContext) => Promise<SlashCommandResult>;
}

export interface SlashCommandContext {
  sessionId?: string;
  model?: string;
  tokenCount?: number;
  costTracker?: CostTracker;
  history?: string[];
  settings?: Record<string, any>;
  onOutput?: (message: string) => void;
}

export interface SlashCommandResult {
  success: boolean;
  message?: string;
  data?: any;
  shouldExit?: boolean;
}

export interface CostTracker {
  totalTokens: number;
  totalCost: number;
  requests: number;
  addUsage(model: string, inputTokens: number, outputTokens: number, latencyMs?: number): any;
  reset(): void;
  getSummary(): string;
}

/**
 * Built-in slash commands
 */
const BUILTIN_COMMANDS: SlashCommand[] = [
  {
    name: 'help',
    aliases: ['h', '?'],
    description: 'Show available commands',
    execute: async (_args, context) => {
      const commands = getRegistry().getAll();
      const lines = ['Available commands:', ''];
      for (const cmd of commands) {
        const aliases = cmd.aliases ? ` (${cmd.aliases.map(a => '/' + a).join(', ')})` : '';
        lines.push(`  /${cmd.name}${aliases} - ${cmd.description}`);
        if (cmd.usage) {
          lines.push(`    Usage: ${cmd.usage}`);
        }
      }
      context.onOutput?.(lines.join('\n'));
      return { success: true, message: lines.join('\n') };
    }
  },
  {
    name: 'cost',
    aliases: ['c'],
    description: 'Show token usage and cost',
    execute: async (_args, context) => {
      if (!context.costTracker) {
        return { success: false, message: 'Cost tracking not enabled' };
      }
      const summary = context.costTracker.getSummary();
      context.onOutput?.(summary);
      return { success: true, message: summary };
    }
  },
  {
    name: 'clear',
    description: 'Clear conversation history',
    execute: async (_args, context) => {
      if (context.history) {
        context.history.length = 0;
      }
      context.onOutput?.('Conversation cleared');
      return { success: true, message: 'Conversation cleared' };
    }
  },
  {
    name: 'model',
    aliases: ['m'],
    description: 'Show or change the current model',
    usage: '/model [model-name]',
    execute: async (args, context) => {
      if (args.length === 0) {
        const msg = `Current model: ${context.model || 'default'}`;
        context.onOutput?.(msg);
        return { success: true, message: msg, data: { model: context.model } };
      }
      const newModel = args[0];
      context.model = newModel;
      const msg = `Model changed to: ${newModel}`;
      context.onOutput?.(msg);
      return { success: true, message: msg, data: { model: newModel } };
    }
  },
  {
    name: 'tokens',
    aliases: ['t'],
    description: 'Show token count for current session',
    execute: async (_args, context) => {
      const count = context.tokenCount || 0;
      const msg = `Token count: ${count}`;
      context.onOutput?.(msg);
      return { success: true, message: msg, data: { tokens: count } };
    }
  },
  {
    name: 'plan',
    aliases: ['p'],
    description: 'Show or manage task plan',
    usage: '/plan [add|remove|clear] [task]',
    execute: async (args, context) => {
      const plans = context.settings?.plans || [];
      
      if (args.length === 0) {
        if (plans.length === 0) {
          context.onOutput?.('No plans set');
          return { success: true, message: 'No plans set' };
        }
        const msg = ['Current plan:', ...plans.map((p: string, i: number) => `  ${i + 1}. ${p}`)].join('\n');
        context.onOutput?.(msg);
        return { success: true, message: msg, data: { plans } };
      }

      const action = args[0];
      const task = args.slice(1).join(' ');

      if (action === 'add' && task) {
        plans.push(task);
        context.settings = { ...context.settings, plans };
        context.onOutput?.(`Added: ${task}`);
        return { success: true, message: `Added: ${task}` };
      }
      if (action === 'remove' && task) {
        const index = parseInt(task) - 1;
        if (index >= 0 && index < plans.length) {
          const removed = plans.splice(index, 1)[0];
          context.settings = { ...context.settings, plans };
          context.onOutput?.(`Removed: ${removed}`);
          return { success: true, message: `Removed: ${removed}` };
        }
      }
      if (action === 'clear') {
        context.settings = { ...context.settings, plans: [] };
        context.onOutput?.('Plans cleared');
        return { success: true, message: 'Plans cleared' };
      }

      return { success: false, message: 'Invalid plan command' };
    }
  },
  {
    name: 'undo',
    aliases: ['u'],
    description: 'Undo last action',
    execute: async (_args, context) => {
      if (context.history && context.history.length > 0) {
        const removed = context.history.pop();
        context.onOutput?.(`Undone: ${removed?.substring(0, 50)}...`);
        return { success: true, message: 'Last action undone' };
      }
      return { success: false, message: 'Nothing to undo' };
    }
  },
  {
    name: 'diff',
    aliases: ['d'],
    description: 'Show pending changes',
    execute: async (_args, context) => {
      const diff = context.settings?.pendingDiff || 'No pending changes';
      context.onOutput?.(diff);
      return { success: true, message: diff };
    }
  },
  {
    name: 'commit',
    description: 'Commit pending changes',
    execute: async (args, context) => {
      const message = args.join(' ') || 'Auto-commit';
      context.onOutput?.(`Committed with message: ${message}`);
      context.settings = { ...context.settings, pendingDiff: null };
      return { success: true, message: `Committed: ${message}` };
    }
  },
  {
    name: 'exit',
    aliases: ['quit', 'q'],
    description: 'Exit the session',
    execute: async (_args, context) => {
      context.onOutput?.('Goodbye!');
      return { success: true, message: 'Exiting', shouldExit: true };
    }
  },
  {
    name: 'settings',
    aliases: ['s'],
    description: 'Show or modify settings',
    usage: '/settings [key] [value]',
    execute: async (args, context) => {
      if (args.length === 0) {
        const settings = JSON.stringify(context.settings || {}, null, 2);
        context.onOutput?.(settings);
        return { success: true, message: settings, data: context.settings };
      }
      if (args.length === 1) {
        const value = context.settings?.[args[0]];
        const msg = `${args[0]}: ${JSON.stringify(value)}`;
        context.onOutput?.(msg);
        return { success: true, message: msg, data: { [args[0]]: value } };
      }
      const [key, ...valueParts] = args;
      let value: any = valueParts.join(' ');
      try {
        value = JSON.parse(value);
      } catch {
        // Keep as string
      }
      context.settings = { ...context.settings, [key]: value };
      context.onOutput?.(`Set ${key} = ${JSON.stringify(value)}`);
      return { success: true, message: `Set ${key}` };
    }
  },
  {
    name: 'map',
    description: 'Show repository map',
    execute: async (_args, context) => {
      const map = context.settings?.repoMap || 'Repository map not available. Run /map refresh to generate.';
      context.onOutput?.(map);
      return { success: true, message: map };
    }
  }
];

/**
 * Slash Command Registry
 */
class SlashCommandRegistry {
  private commands: Map<string, SlashCommand> = new Map();
  private aliases: Map<string, string> = new Map();

  constructor() {
    // Register built-in commands
    for (const cmd of BUILTIN_COMMANDS) {
      this.register(cmd);
    }
  }

  register(command: SlashCommand): void {
    this.commands.set(command.name, command);
    if (command.aliases) {
      for (const alias of command.aliases) {
        this.aliases.set(alias, command.name);
      }
    }
  }

  get(name: string): SlashCommand | undefined {
    const resolved = this.aliases.get(name) || name;
    return this.commands.get(resolved);
  }

  getAll(): SlashCommand[] {
    return Array.from(this.commands.values());
  }

  has(name: string): boolean {
    return this.commands.has(name) || this.aliases.has(name);
  }
}

let registry: SlashCommandRegistry | null = null;

export function getRegistry(): SlashCommandRegistry {
  if (!registry) {
    registry = new SlashCommandRegistry();
  }
  return registry;
}

export function registerCommand(command: SlashCommand): void {
  getRegistry().register(command);
}

/**
 * Parse slash command from input
 */
export function parseSlashCommand(input: string): { command: string; args: string[] } | null {
  const trimmed = input.trim();
  if (!trimmed.startsWith('/')) {
    return null;
  }

  const parts = trimmed.slice(1).split(/\s+/);
  const command = parts[0].toLowerCase();
  const args = parts.slice(1);

  return { command, args };
}

/**
 * Execute a slash command
 */
export async function executeSlashCommand(
  input: string,
  context: SlashCommandContext
): Promise<SlashCommandResult> {
  const parsed = parseSlashCommand(input);
  if (!parsed) {
    return { success: false, message: 'Not a slash command' };
  }

  const command = getRegistry().get(parsed.command);
  if (!command) {
    return { success: false, message: `Unknown command: /${parsed.command}` };
  }

  return command.execute(parsed.args, context);
}

/**
 * Check if input is a slash command
 */
export function isSlashCommand(input: string): boolean {
  return input.trim().startsWith('/');
}

/**
 * Slash Command Handler for CLI integration
 */
export class SlashCommandHandler {
  private context: SlashCommandContext;

  constructor(initialContext: Partial<SlashCommandContext> = {}) {
    this.context = {
      history: [],
      settings: {},
      ...initialContext
    };
  }

  async handle(input: string): Promise<SlashCommandResult> {
    return executeSlashCommand(input, this.context);
  }

  isCommand(input: string): boolean {
    return isSlashCommand(input);
  }

  getContext(): SlashCommandContext {
    return this.context;
  }

  updateContext(updates: Partial<SlashCommandContext>): void {
    this.context = { ...this.context, ...updates };
  }

  registerCommand(command: SlashCommand): void {
    registerCommand(command);
  }
}

export function createSlashCommandHandler(context?: Partial<SlashCommandContext>): SlashCommandHandler {
  return new SlashCommandHandler(context);
}
